### PR TITLE
Preserve content with Get-AuthenticodeSignature

### LINKED ISSUE #18773: Get-AuthenticodeSignature Displaying Inconsistent Behavior Between -FilePath and -Content
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

I am currently experiencing an issue with `Get-AuthenticodeSignature` where when checking the signature with a direct file path the signature is valid, however when converting the same file into a byte array and attempting to run `Get-AuthenticodeSignature` against it, it reports no valid signature.

```powershell
Get-AuthenticodeSignature -FilePath "~\Repos\Strapper\Strapper\Strapper.psm1"
<#
SignerCertificate                         Status               StatusMessage        Path
-----------------                         ------               -------------        ----
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF  Valid                Signature verified.  Strapper.psm1
#>


Get-AuthenticodeSignature -Content (Get-Content -Path "~\Repos\Strapper\Strapper\Strapper.psm1" -AsByteStream) -SourcePathOrExtension psm1
<#
SignerCertificate                         Status               StatusMessage        Path
-----------------                         ------               -------------        ----
                                          NotSigned            The file psm1 is no… psm1
#>
```
I assumed that this could have been a file extension issue, and it does appear to be replicated with even simple files, but changing the extension doesn't appear to fix the issue:

```powershell
Get-AuthenticodeSignature -FilePath "~\Downloads\Get-Host.psm1"
<#
SignerCertificate                         Status               StatusMessage        Path
-----------------                         ------               -------------        ----
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF  Valid                Signature verified.  Get-Host.psm1
#>

Get-AuthenticodeSignature -Content (Get-Content "~\Downloads\Get-Host.psm1" -AsByteStream) -SourcePathOrExtension psm1
<#
SignerCertificate                         Status               StatusMessage        Path
-----------------                         ------               -------------        ----
                                          NotSigned            The file psm1 is no… psm1
#>

Get-AuthenticodeSignature -FilePath "~\Downloads\Get-Host.ps1"
<#
SignerCertificate                         Status               StatusMessage        Path
-----------------                         ------               -------------        ----
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF  Valid                Signature verified.  Get-Host.ps1
#>

Get-AuthenticodeSignature -Content (Get-Content "~\Downloads\Get-Host.ps1" -AsByteStream) -SourcePathOrExtension ps1
<#
SignerCertificate                         Status               StatusMessage        Path
-----------------                         ------               -------------        ----
                                          NotSigned            The file ps1 is not… ps1
#>
```
## `Get-Host.psm1`
```powershell
Get-Host
# SIG # Begin signature block
# MIInbwYJKoZIhvcN...
# SIG # End signature block
```

### Expected behavior

```console
PS> Get-AuthenticodeSignature -Content (Get-Content -Path "~\Repos\Strapper\Strapper\Strapper.psm1" -AsByteStream) -SourcePathOrExtension psm1
SignerCertificate                         Status               StatusMessage        Path
-----------------                         ------               -------------        ----
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF  Valid                Signature verified.  Strapper.psm1
```

### Actual behavior

```console
PS> Get-AuthenticodeSignature -Content (Get-Content -Path "~\Repos\Strapper\Strapper\Strapper.psm1" -AsByteStream) -SourcePathOrExtension psm1
SignerCertificate                         Status               StatusMessage        Path
-----------------                         ------               -------------        ----
                                          NotSigned            The file psm1 is no… psm1
```

### Error details

_No response_

### Environment data

```powershell
PS> $PSVersionTable
Name                           Value
----                           -----
PSVersion                      7.3.0
PSEdition                      Core
GitCommitId                    7.3.0
OS                             Microsoft Windows 10.0.22621
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

image

### PR DESCRIPTION
**Pull request by jborean93**
# PR Summary

Do not try and roundtrip the `-Content` bytes to a Unicode encoded string and then back to bytes. Instead just use the raw input bytes when validating the content signature.

## PR Context

Fixes https://github.com/PowerShell/PowerShell/issues/18773

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Review by iSazonov — APPROVED** [review]

**Comment by jborean93** [comment]
Any updates on this one, have been a few months with CI being green.

File: src/System.Management.Automation/security/SecurityManager.cs [hunk 1]
```diff
@@ -535,15 +535,15 @@ private static Signature GetSignatureWithEncodingRetry(string path, ExternalScri
             // try harder to validate the signature by being explicit about encoding
             // and providing the script contents
             string verificationContents = Encoding.Unicode.GetString(script.OriginalEncoding.GetPreamble()) + script.ScriptContents;
-            signature = SignatureHelper.GetSignature(path, verificationContents);
+            signature = SignatureHelper.GetSignature(path, Encoding.Unicode.GetBytes(verificationContents));
```
Comment (daxian-dbw) [inline src/System.Management.Automation/security/SecurityManager.cs:538]: The existing code seems wrong here -- the `OriginalEncoding` could be an encoding other than `Unicode`, and that means `Encoding.Unicode.GetBytes(verificationContents)` could be different byte sequences than the original bytes from the file.
I will make a change and see if all tests still pass.

**Review by daxian-dbw — APPROVED** [review]
LGTM.
@SeeminglyScience I made a change on how we get the byte sequences for the content of a script file because the existing code seems incorrect to me. CIs still passed with my changes, but I'd like you to take a look just to make sure the changes make sense. Thanks!

**Comment by SeeminglyScience** [comment]
> LGTM. @SeeminglyScience I made a change on how we get the byte sequences for the content of a script file because the existing code seems incorrect to me. CIs still passed with my changes, but I'd like you to take a look just to make sure the changes make sense. Thanks!

Yeah I think your changes make sense. It may actually solve some finickyness with encoding requirements and signing that I vaguely remember hearing about.  LGTM!

File: src/System.Management.Automation/security/SecurityManager.cs [hunk 2]
```diff
@@ -552,6 +552,17 @@ private static Signature GetSignatureWithEncodingRetry(string path, ExternalScri
             return signature;
         }
 
+        private static byte[] GetContentBytesWithBom(Encoding encoding, string scriptContent)
+        {
+            ReadOnlySpan<byte> bomBytes = encoding.Preamble;
+            byte[] contentBytes = encoding.GetBytes(scriptContent);
```
[the remark this answers is not in the document]
Comment (daxian-dbw) [reply src/System.Management.Automation/security/SecurityManager.cs:558]: It's been using `System.Text.Encoding.Unicode.GetBytes(fileContent);` in `GetWinTrustData` method, so I guess the LOH allocation has not become a concern, but yes, it's a potential perf issue.

**Pull request merged**